subsys/net/lib/dfu_client/src/http_dfu_client.c

```diff
@@ -0,0 +1,278 @@
+/*
+ * Copyright (c) 2018 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <http_client.h>
+#include <dfu_client.h>
+#include <net/socket.h>
+#include <zephyr/types.h>
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(http_dfu);
+
+#define REQUEST_TEMPLATE "GET %s HTTP/1.1\r\n"\
+	"Host: %s\r\n"\
+	"Connection: keep-alive\r\n"\
+	"Range: bytes=%d-%d\r\n\r\n"
+
+
+static void rxdata_flush(struct dfu_client_object * const dfu)
+{
+	if (dfu == NULL || dfu->host == NULL || dfu->callback == NULL || dfu->status == DFU_CLIENT_STATUS_IDLE) {
+		return;
+	}
+
+	int flush_len = httpc_recv(dfu->fd, dfu->resp_buf, CONFIG_NRF_DFU_HTTP_MAX_RESPONSE_SIZE, 0);
+
+	LOG_INF("rxdata_flush, len %d\n", flush_len);
+	if (flush_len == -1) {
+		if (errno == EAGAIN) {
+			LOG_ERR("flused %d\n", flush_len);
+		} else {
+			/* Something is wrong on the socket! */
+			dfu->callback(dfu, DFU_CLIENT_EVT_ERROR, errno);
+		}
+	}
+
+	memset(dfu->resp_buf, 0, CONFIG_NRF_DFU_HTTP_MAX_RESPONSE_SIZE);
+
+}
+
+
+static int fragment_request(struct dfu_client_object * const dfu, bool flush, bool connnection_close)
+{
+	if (dfu == NULL || dfu->host == NULL || dfu->callback == NULL || dfu->resource == NULL) {
+		LOG_ERR("fragment_request(): Invalid dfu object");
+		return -1;
+	}
+
+	if (flush == true) {
+		rxdata_flush(dfu);
+	}
+
+	if (connnection_close == true) {
+		httpc_disconnect(dfu->fd);
+		dfu->fd = -1;
+		dfu->fd = httpc_connect(dfu->host, NULL, AF_INET, IPPROTO_TCP);
+		if (dfu->fd < 0) {
+			LOG_ERR("fragment_request(): httpc_connect() failed, err %d", errno);
+			dfu->callback(dfu, DFU_CLIENT_EVT_ERROR, ECONNRESET);
+			return -1;
+		}
+	}
+
+	memset(dfu->req_buf, 0, CONFIG_NRF_DFU_HTTP_MAX_REQUEST_SIZE);
+
+	int request_len = snprintf(dfu->req_buf, CONFIG_NRF_DFU_HTTP_MAX_REQUEST_SIZE, REQUEST_TEMPLATE, dfu->resource,
+							dfu->host, dfu->download_size,
+							(dfu->download_size + CONFIG_NRF_DFU_HTTP_MAX_FRAGMENT_SIZE - 1));
+
+	LOG_INF("fragment_request(), request length %d, state = %d\n", request_len, dfu->status);
+
+	if (request_len > 0) {
```

Comment:
Why do you want to send a partial request?

Reply:
I don't, that's the whole point. With current implementation, if certain combination of `dfu->resource`, `dfu->host` etc. does not fit into preallocated buffer, it will get silently truncated and an incomplete request will be sent to the server. It's better to return an error to the application in this case.

Reply:
For the use case you describe snprintf helps us detect this. A return value equal or greater than the indicated size implies truncated buffer according to the man page. 

However, send does not guarantee that all that we request to send was sent. So, added a check for this.

If you have specific ideas in mind, please suggest.

Reply:
Yep, `snprintf` does the job for us, all we need to do is check, thanks for adding that.

As for the `send`, a loop + offset tracking typically does the job:
https://github.com/zephyrproject-rtos/zephyr/blob/b87920bf3c789ad4eeebb749d9a06ae72f15d589/subsys/net/lib/mqtt/mqtt_transport_socket_tcp.c#L71

Reply:
For HTTP, sending an incomplete request can be an attack vector. I am sure you have heard of the Slow Loris attack. Here the server is expected to wait for the complete request to arrive.  Note that we use the connection header with keep-alive in our request.

I can add the loop you suggest. But failure to finish despite the loop is still an issue.

The best behavior I would suggest is to close the connection and start again. At least on cellular devices.

For MQTT, I can see a similar concern on a having sent partial request and then exiting with error. There is never a guarantee that subsequent requests will succeed. There is not even better probability for success just because we use a loop.
 
So, I would say an error event to request closing of connection from the application.
This is a suggestion. I will make the change you decide.

Reply:
This library seems to notify such errors to the application, I'd be uniform here and do the same in case of `send` error.

As for the MQTT, the library closes TCP connection in case of `send` error.